lahja/asyncio/endpoint.py

```diff
@@ -402,21 +485,29 @@ def run(self, *args, **kwargs):  # type: ignore
     async def _accept_conn(self, reader: StreamReader, writer: StreamWriter) -> None:
         conn = Connection(reader, writer)
         remote = RemoteEndpoint(None, conn, self._receiving_queue.put)
-        self._half_connections.add(remote)
 
         task = asyncio.ensure_future(self._handle_client(remote))
         task.add_done_callback(self._server_tasks.remove)
+        task.add_done_callback(lambda _: self._half_connections.remove(remote))
         self._server_tasks.add(task)
 
-        # the Endpoint on the other end blocks until it receives this message
-        await remote.notify_subscriptions_updated(self.subscribed_events)
+        await remote.wait_started()
+
+        # we **must** ensure that the subscription updates are locked between
+        # the time that we manually update this individual connection and that
+        # we place it within the set of tracked connections, otherwise, a
+        # subscription update from elsewhere can occur between the time these
+        # two statements execute resulting in the remote missing a new
+        # subscription update.  Note that inverting these statements should
+        # also mitigate this, but it has the downside of the manual update
+        # potentially being redundant.
+        async with self._subscription_updates_condition:
+            await remote.notify_subscriptions_updated(self.subscribed_events)
+            await self._add_half_connection(remote)
```

Comment:
Ahh, okay, I see the problem now, `_add_half_connection` blocks when it tries to acquire `connections_changed` so there's a period in time where the remote isn't receiving subscription updates. Double-sending the same set of subscriptions has a very small cost though! And I think that trying not to send it introduced greater inefficiencies elsewhere, serializing all updates isn't cheap!

Reply:
(I thought I responded to this somewhere but I don't see it)

I'm inclined to leave it this way because it *correctly* manages the appropriate locks on the resource, where-as the other mechanism solves the problem through indirection.  First, I think the overall cost here is low because our usage patterns for the library don't involve adding new connections at a high frequency.  Second, doing it the alternate way of inverting these statements seems like a solution that could easily be undone incidentally in a refactor since it's an implicit solution.